crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -1847,16 +1794,18 @@ where
             // we're pending if both engine messages and sync events are pending (fully drained)
             let is_pending = engine_messages_pending && sync_pending;
 
-            // Poll prune controller if all conditions are met:
-            // 1. Pipeline is idle
-            // 2. No engine and sync messages are pending
-            // 3. Latest FCU status is not INVALID
-            if this.sync.is_pipeline_idle() &&
-                is_pending &&
-                !this.forkchoice_state_tracker.is_latest_invalid()
-            {
-                if let Some(res) = this.poll_prune(cx) {
-                    return Poll::Ready(res)
+            // Poll next hook if all conditions are met:
+            // 1. No engine and sync messages are pending
+            // 2. Latest FCU status is not INVALID
+            if is_pending && !this.forkchoice_state_tracker.is_latest_invalid() {
+                if let Poll::Ready(result) = this.hooks.poll_next_hook(
+                    cx,
+                    HookArguments { tip_block_number: this.blockchain.canonical_tip().number },
+                    HookDependencies { db_write: this.sync.is_pipeline_active() },
+                ) {
+                    if let Err(err) = this.on_hook_action(result?) {
+                        return Poll::Ready(Err(err))
+                    }
                 }
             }
```

Comment:
So we are only polling one hook (maybe two if there was a previous write poll)  per engine hook. 

Is that intended? why not more? we don't want to potentially block the FCU handle?

Reply:
could we run them in parallel here?

Reply:
> we don't want to potentially block the FCU handle?

yes, that was the main reason, so we would return the control asap and process any new incoming messages, because they're higher priority than hooks